testnet/stacks-node/src/run_loop/puppet.rs

```diff
@@ -0,0 +1,176 @@
+use crate::ChainTip;
+use async_std::net::TcpListener;
+use async_std::stream::StreamExt;
+use async_std::task::block_on;
+use http_types::{Method, Response, StatusCode};
+use std::borrow::BorrowMut;
+use std::io;
+use std::io::{Error, Write};
+use std::ops::Add;
+use std::sync::{Arc, Mutex};
+use std::thread::{sleep, Builder, JoinHandle};
+use std::time::{Duration, SystemTime};
+
+struct PuppetControl {
+    pub current_block: u64,
+    pub target_block: u64,
+    pub next_block_time: SystemTime,
+    pub block_interval: Duration,
+}
+
+impl Default for PuppetControl {
+    fn default() -> Self {
+        let default_interval = Duration::from_secs(600);
+        Self {
+            current_block: 0,
+            target_block: 1,
+            next_block_time: SystemTime::now().add(default_interval),
+            block_interval: default_interval,
+        }
+    }
+}
+
+pub struct PuppetController {
+    inner: Arc<Mutex<PuppetControl>>,
+    bind_addr: String,
+    join_handle: Option<JoinHandle<Result<(), Error>>>,
+}
+
+impl PuppetController {
+    pub fn new(bind_addr: &str) -> Self {
+        Self {
+            inner: Arc::new(Mutex::new(PuppetControl::default())),
+            bind_addr: bind_addr.to_string(),
+            join_handle: None,
+        }
+    }
+
+    pub fn start(&mut self) {
+        if self.join_handle.is_some() {
+            warn!("Puppet mode control server is already started");
+            return ();
+        }
+        info!("Starting puppet mode control server..");
+        let puppet_control = Arc::clone(&self.inner);
+        let puppet_bind = self.bind_addr.to_string();
+        self.join_handle = Some(
+            Builder::new()
+                .name("puppet".into())
+                .spawn(move || {
+                    block_on(async {
+                        let listener = TcpListener::bind(puppet_bind).await?;
+                        info!(
+                            "Start puppet mode control server on: {}",
+                            listener.local_addr()?
+                        );
+
+                        // For each incoming TCP connection, spawn a task and call `accept`.
+                        let mut incoming = listener.incoming();
+                        while let Some(stream) = incoming.next().await {
+                            if stream.is_err() {
+                                return Err(stream.unwrap_err());
+                            }
+                            let stream = stream?;
+                            async_h1::accept(stream.clone(), |req| async {
+                                let mut req = req;
+                                match (req.method(), req.url().path()) {
+                                    (Method::Get, "/ping") => Ok(Response::new(StatusCode::Ok)),
```

Comment:
Could you prefix the API with both a system identifier (e.g. `puppet`) and a version (e.g. `v1`)?  So, this would be `/puppet/v1/ping`.  Thanks!

Reply:
Prefix added. But I don't understand why adding a system identifier is necessary. It does make sense if we mix the API into a currently existing one, but we're not.